### PR TITLE
Part 3 of 4: Auditing our required parameters for the C# SDK. 

### PR DESCRIPTION
**Pull request by timothodge**
Looking to fix things up based on this open issue: https://github.com/smartsheet/smartsheet-csharp-sdk/issues/25

This is 3/4 of the audit done. Here is the public API docs for anyone who wants to cross check: https://smartsheet.redoc.ly/tag/groups/

### CODE REVIEW

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetAttachmentResources.cs [hunk 1]
```diff
@@ -45,7 +45,7 @@ public interface SheetAttachmentResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        Attachment AttachFile(long sheetId, string file, string fileType);
+        Attachment AttachFile(long sheetId, string file, string? fileType = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetAttachmentResources.cs:48]: Screenshot 2024-04-30 at 4 17 11 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetAttachmentResources.cs [hunk 2]
```diff
@@ -115,7 +115,7 @@ public interface SheetAttachmentResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<Attachment> ListAttachments(long sheetId, PaginationParameters paging);
+        PaginatedResult<Attachment> ListAttachments(long sheetId, PaginationParameters? paging = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetAttachmentResources.cs:118]: Screenshot 2024-04-30 at 4 17 40 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetAutomationRuleResources.cs [hunk 3]
```diff
@@ -44,7 +44,7 @@ public interface SheetAutomationRuleResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<AutomationRule> ListAutomationRules(long sheetId, PaginationParameters paging);
+        PaginatedResult<AutomationRule> ListAutomationRules(long sheetId, PaginationParameters? paging = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetAutomationRuleResources.cs:47]: Screenshot 2024-04-30 at 4 20 54 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetColumnResources.cs [hunk 4]
```diff
@@ -45,7 +45,7 @@ public interface SheetColumnResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<Column> ListColumns(long sheetId, IEnumerable<ColumnInclusion> include, PaginationParameters paging);
+        PaginatedResult<Column> ListColumns(long sheetId, IEnumerable<ColumnInclusion>? include = null, PaginationParameters? paging = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetColumnResources.cs:48]: Screenshot 2024-04-30 at 4 27 23 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetColumnResources.cs [hunk 5]
```diff
@@ -112,7 +112,7 @@ public interface SheetColumnResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        Column GetColumn(long sheetId, long columnId, IEnumerable<ColumnInclusion> include);
+        Column GetColumn(long sheetId, long columnId, IEnumerable<ColumnInclusion>? include = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetColumnResources.cs:115]: Screenshot 2024-04-30 at 4 29 28 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetCrossSheetReferenceResources.cs [hunk 6]
```diff
@@ -42,7 +42,7 @@ public interface SheetCrossSheetReferenceResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<CrossSheetReference> ListCrossSheetReferences(long sheetId, PaginationParameters paging);
+        PaginatedResult<CrossSheetReference> ListCrossSheetReferences(long sheetId, PaginationParameters? paging = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetCrossSheetReferenceResources.cs:45]: Screenshot 2024-04-30 at 4 37 13 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetDiscussionResources.cs [hunk 7]
```diff
@@ -89,7 +89,7 @@ public interface SheetDiscussionResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<Discussion> ListDiscussions(long sheetId, IEnumerable<DiscussionInclusion> include, PaginationParameters paging);
+        PaginatedResult<Discussion> ListDiscussions(long sheetId, IEnumerable<DiscussionInclusion>? include = null, PaginationParameters? paging = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetDiscussionResources.cs:92]: Screenshot 2024-04-30 at 4 41 12 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs [hunk 8]
```diff
@@ -59,7 +59,7 @@ public interface SheetResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<Sheet> ListSheets(IEnumerable<SheetInclusion> includes, PaginationParameters paging, DateTime? modifiedSince = null);
+        PaginatedResult<Sheet> ListSheets(IEnumerable<SheetInclusion>? includes = null, PaginationParameters? paging = null, DateTime? modifiedSince = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs:62]: Screenshot 2024-04-30 at 4 51 04 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs [hunk 9]
```diff
@@ -76,7 +76,7 @@ public interface SheetResources
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
         [Obsolete("use Smartsheet.UserResources.SheetResources.ListOrgSheets", true)]
         [EditorBrowsable(EditorBrowsableState.Never)]
-        PaginatedResult<Sheet> ListOrganizationSheets(PaginationParameters paging);
+        PaginatedResult<Sheet> ListOrganizationSheets(PaginationParameters? paging = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs:79]: Screenshot 2024-05-01 at 11 12 13 AM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs [hunk 10]
```diff
@@ -101,13 +101,13 @@ public interface SheetResources
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
         Sheet GetSheet(
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs:102]: Screenshot 2024-05-01 at 11 12 53 AM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs [hunk 11]
```diff
@@ -208,7 +208,7 @@ public interface SheetResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        void GetSheetAsPDF(long sheetId, BinaryWriter outputStream, PaperSize? paperSize);
+        void GetSheetAsPDF(long sheetId, BinaryWriter outputStream, PaperSize? paperSize = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs:211]: Same as above GetSheet look at possible alternatives.

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs [hunk 12]
```diff
@@ -404,23 +404,6 @@ public interface SheetResources
...
-        /// <param name="destination"> the destination to copy to </param>
-        /// <param name="include"> the elements to copy. Note: Cell history will not be copied, regardless of which include parameter values are specified.</param>
-        /// <returns> the created folder </returns>
-        /// <exception cref="System.InvalidOperationException"> if any argument is null or an empty string </exception>
-        /// <exception cref="InvalidRequestException"> if there is any problem with the REST API request </exception>
-        /// <exception cref="AuthorizationException"> if there is any problem with the REST API authorization (access token) </exception>
-        /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
-        /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
-        /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        Sheet CopySheet(long sheetId, ContainerDestination destination, IEnumerable<SheetCopyInclusion> include);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs:422]: We can remove this now that includes and excludes are optional below.

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs [hunk 13]
```diff
@@ -437,7 +420,7 @@ public interface SheetResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        Sheet CopySheet(long sheetId, ContainerDestination destination, IEnumerable<SheetCopyInclusion> include, IEnumerable<SheetCopyExclusion> exclude);
+        Sheet CopySheet(long sheetId, ContainerDestination destination, IEnumerable<SheetCopyInclusion>? include = null, IEnumerable<SheetCopyExclusion>? exclude = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs:423]: Screenshot 2024-05-01 at 11 17 30 AM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs [hunk 14]
```diff
@@ -455,22 +438,6 @@ public interface SheetResources
...
-        /// <param name="sheetId"> the sheet Id </param>
-        /// <param name="sortSpecifier"> the sort criteria </param>
-        /// <returns> the Sheet (note that if there is no such resource, this method will throw a ResourceNotFoundException rather than returning null). </returns>
-        /// <exception cref="System.InvalidOperationException"> if any argument is null or an empty string </exception>
-        /// <exception cref="InvalidRequestException"> if there is any problem with the REST API request </exception>
-        /// <exception cref="AuthorizationException"> if there is any problem with the REST API authorization (access token) </exception>
-        /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
-        /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
-        /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        Sheet SortSheet(long sheetId, SortSpecifier sortSpecifier);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs:472]: Can be removed now that level is optional below.

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs [hunk 15]
```diff
@@ -486,7 +453,7 @@ public interface SheetResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        Sheet SortSheet(long sheetId, SortSpecifier sortSpecifier, int? level);
+        Sheet SortSheet(long sheetId, SortSpecifier sortSpecifier, int? level= null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs:456]: Screenshot 2024-05-01 at 11 18 46 AM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs [hunk 16]
```diff
@@ -503,7 +470,7 @@ public interface SheetResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        Sheet ImportCsvSheet(string file, string sheetName, int? headerRowIndex, int? primaryColumnIndex);
+        Sheet ImportCsvSheet(string file, string? sheetName = null, int? headerRowIndex = null, int? primaryColumnIndex = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs:473]: Screenshot 2024-05-01 at 11 20 19 AM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetRowResources.cs [hunk 17]
```diff
@@ -80,7 +80,7 @@ public interface SheetRowResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        Row GetRow(long sheetId, long rowId, IEnumerable<RowInclusion> include, IEnumerable<RowExclusion> exclude);
+        Row GetRow(long sheetId, long rowId, IEnumerable<RowInclusion>? include = null, IEnumerable<RowExclusion>? exclude = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetRowResources.cs:83]: Screenshot 2024-05-01 at 11 27 54 AM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetRowResources.cs [hunk 18]
```diff
@@ -102,7 +102,7 @@ public interface SheetRowResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        CopyOrMoveRowResult CopyRowsToAnotherSheet(long sheetId, CopyOrMoveRowDirective directive, IEnumerable<CopyRowInclusion> include, bool? ignoreRowsNotFound);
+        CopyOrMoveRowResult CopyRowsToAnotherSheet(long sheetId, CopyOrMoveRowDirective directive, IEnumerable<CopyRowInclusion>? include = null, bool? ignoreRowsNotFound = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetRowResources.cs:105]: Screenshot 2024-05-01 at 11 29 01 AM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetRowResources.cs [hunk 19]
```diff
@@ -120,7 +120,7 @@ public interface SheetRowResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        IList<long> DeleteRows(long sheetId, IEnumerable<long> ids, bool? ignoreRowsNotFound);
+        IList<long> DeleteRows(long sheetId, IEnumerable<long> ids, bool? ignoreRowsNotFound = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetRowResources.cs:123]: Screenshot 2024-05-01 at 11 29 51 AM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetRowResources.cs [hunk 20]
```diff
@@ -143,7 +143,7 @@ public interface SheetRowResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        CopyOrMoveRowResult MoveRowsToAnotherSheet(long sheetId, CopyOrMoveRowDirective directive, IEnumerable<MoveRowInclusion> include, bool? ignoreRowsNotFound);
+        CopyOrMoveRowResult MoveRowsToAnotherSheet(long sheetId, CopyOrMoveRowDirective directive, IEnumerable<MoveRowInclusion>? include = null, bool? ignoreRowsNotFound = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetRowResources.cs:146]: Screenshot 2024-05-01 at 11 30 40 AM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetSummaryResources.cs [hunk 21]
```diff
@@ -43,7 +43,7 @@ public interface SheetSummaryResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        SheetSummary GetSheetSummary(long sheetId, IEnumerable<SummaryFieldInclusion> include, IEnumerable<SummaryFieldExclusion> exclude);
+        SheetSummary GetSheetSummary(long sheetId, IEnumerable<SummaryFieldInclusion>? include = null, IEnumerable<SummaryFieldExclusion>? exclude = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetSummaryResources.cs:46]: Screenshot 2024-05-01 at 11 40 18 AM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetSummaryResources.cs [hunk 22]
```diff
@@ -60,7 +60,7 @@ public interface SheetSummaryResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<SummaryField> GetSheetSummaryFields(long sheetId, IEnumerable<SummaryFieldInclusion> include, IEnumerable<SummaryFieldExclusion> exclude, PaginationParameters paging);
+        PaginatedResult<SummaryField> GetSheetSummaryFields(long sheetId, IEnumerable<SummaryFieldInclusion>? include = null, IEnumerable<SummaryFieldExclusion>? exclude = null, PaginationParameters? paging = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetSummaryResources.cs:63]: Screenshot 2024-05-01 at 11 40 43 AM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetSummaryResources.cs [hunk 23]
```diff
@@ -76,7 +76,7 @@ public interface SheetSummaryResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        IList<SummaryField> AddSheetSummaryFields(long sheetId, IEnumerable<SummaryField> fields, bool? renameIfConflict);
+        IList<SummaryField> AddSheetSummaryFields(long sheetId, IEnumerable<SummaryField> fields, bool? renameIfConflict = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetSummaryResources.cs:79]: Screenshot 2024-05-01 at 11 41 15 AM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetSummaryResources.cs [hunk 24]
```diff
@@ -140,7 +140,7 @@ public interface SheetSummaryResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        IList<long> DeleteSheetSummaryFields(long sheetId, IEnumerable<long> fieldIds, bool? ignoreSummaryFieldsNotFound);
+        IList<long> DeleteSheetSummaryFields(long sheetId, IEnumerable<long> fieldIds, bool? ignoreSummaryFieldsNotFound = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetSummaryResources.cs:143]: Screenshot 2024-05-01 at 11 42 07 AM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/SheetSummaryResources.cs [hunk 25]
```diff
@@ -158,6 +158,6 @@ public interface SheetSummaryResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        SummaryField AddSheetSummaryFieldImage(long sheetId, long fieldId, string file, string fileType, string altText);
+        SummaryField AddSheetSummaryFieldImage(long sheetId, long fieldId, string file, string? fileType = null, string? altText = null);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/SheetSummaryResources.cs:161]: Screenshot 2024-05-01 at 11 42 40 AM

**Review by rzaracota — APPROVED** [review]
This makes the API look so much nicer, thank you!

**Pull request merged**